frametrim/ft_d3d11.cpp

```diff
@@ -0,0 +1,1835 @@
+{
+    return width_stride;
+}
+
+D3D11Texture2D::D3D11Texture2D(ImplPtr impl, void *id, unsigned height, unsigned bind_flags):
+    D3D11Resource(impl, id),
+    m_height(height),
+    m_bind_flags(bind_flags)
+{
+}
+
+unsigned
+D3D11Texture2D::getMappedSize(unsigned width_stride, unsigned depth_stride)
+{
+    return m_height * width_stride;
+}
+
+D3D11Texture3D::D3D11Texture3D(ImplPtr impl, void *id, unsigned depth):
+    D3D11Resource(impl, id),
+    m_depth(depth)
+{
+}
+
+unsigned
+D3D11Texture3D::getMappedSize(unsigned width_stride, unsigned depth_stride)
+{
+    return m_depth * depth_stride;
+}
+
+D3D11View::D3D11View(ImplPtr impl, void *id, D3D11Resource::Pointer res):
+    D3D11DeviceChild(impl, id),
+    m_resource(res)
+{
+    addDependency(res);
+}
+
+void
+D3D11View::addUpdateCall(const trace::Call& call)
+{
+    m_resource->addCall(getInitCall());
+    m_resource->addUpdateCall(call);
+}
+
+void
+D3D11View::addUpdateDependency(Object::Pointer dep)
+{
+    m_resource->addUpdateDependency(dep);
+}
+
+void
+D3D11View::clear(const trace::Call& call)
+{
+    m_resource->clear();
+    m_resource->addCall(getInitCall());
+    m_resource->addUpdateCall(call);
+}
+
+D3D11Context::D3D11Context(ImplPtr impl, void *id, bool deferred):
+    D3D11DeviceChild(impl, id),
+    m_deferred(deferred)
+{
+}
+
+void
+D3D11Context::bindObject(const trace::Call& call, ePerContextBinding binding_type,
+                         unsigned bindpoint, unsigned slot, void *bound_obj_id,
+                         const trace::Array *deps=nullptr)
+{
+    auto& bindings = m_bindings[binding_type];
+    auto bound_obj = lookup<D3D11DeviceChild>(bound_obj_id);
+
+    // TODO cleanup
+    auto binding = fakeCreate<D3D11Binding>(call, bound_obj);
+    binding->addDependencies(deps);
+    bindings[bindpoint][slot] = binding;
+
+    if (bound_obj_id)
+        assert(bound_obj);
+
+    /* Make sure the bound object and dependencies initialization is recorderd */
+    if (bound_obj)
+        m_impl.lock()->recordObjectInit(bound_obj);
+    if (deps) {
+        for (auto dep: deps->values) {
+            auto obj = lookup<Object>(dep->toPointer());
+            m_impl.lock()->recordObjectInit(obj);
+        }
+    }
+}
+
+void
+D3D11Context::bindObjects(const trace::Call& call, ePerContextBinding binding_type,
+                          unsigned bindpoint, unsigned slot, unsigned param_id)
+{
+    auto objects = call.arg(param_id).toArray();
+
+    // TODO clear all current bindings
+    if (!objects)
+        return;
+
+    for (auto object: objects->values) {
+        bindObject(call, binding_type, bindpoint, slot, object->toPointer());
+        slot++;
+    }
+}
+
+void
+D3D11Context::setObjects(const trace::Call& call, ePerContextBinding binding_type,
+                         unsigned bindpoint, unsigned slot, unsigned param_id)
+{
+    auto& bindings = m_bindings[binding_type];
+    bindings.clear();
```

Comment:
It is already the case. `setObjects` is only used for `OMSetRenderTargets` and `BindRenderTargets{AndUAVS}`. The other methods use `bindObjects`, `bindSlots`, etc.

Would you prefer to have a new argument instead of a different method to differentiate the behaviors?

Reply:
It's fine as is then.  Let's just add a comment on each function explaining the ref count semantics.